library_management_system_downloader/run_api_download_and_parse_results.py

```diff
@@ -0,0 +1,210 @@
+
+dois_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+library_holdings_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+# Create an sqlalchemy portal for inserting new records into
+# the tables:
+dois_table_insert = dois_table.insert()
+holdings_table_insert = library_holdings_table.insert()
+
+# =============================================================================
+# Define functions for querying and inserting into the database
+# =============================================================================
+
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+
+def is_doi_already_answered_in_database(
+        doi,
+        sqlalchemy_session=sql_session):
+    existing_doi_with_fulltext_answer = sql_session.query(
+            distinct(dois_table.c.doi)) \
+            .join(library_holdings_table) \
+            .filter(dois_table.c.doi == doi) \
+            .filter(library_holdings_table.c.full_text_indicator.isnot(None)) \
+            .first()
+
+    return bool(existing_doi_with_fulltext_answer)  # Returns False or True.
+
+# Example invocation:
+# is_doi_already_answered_in_database('10.1001/2012.jama.10139')
+
+
+def insert_a_doi_database_record(
+        doi,
+        api_response_text,
+        full_text_indicator_for_doi,
+        sqlalchemy_session=sql_session):
+    doi_foreign_key = sql_session.query(
+            distinct(dois_table.c.database_id)).filter(
+                    dois_table.c.doi == doi).first()
+
+    if doi_foreign_key is None:
+        logging.info(f'Inserting DOI "{doi}" into the database...')
+        inserted_doi = dois_table_insert.execute(doi=doi)
+        # Get the foreign key we just inserted into the database:
+        doi_foreign_key = inserted_doi.inserted_primary_key[0]
+    else:
+        doi_foreign_key = doi_foreign_key[0]
+        logging.info(f'DOI "{doi}" is already in the database (foreign key '
+                     '"{doi_foreign_key}"), so not inserting it...')
+
+    # Insert the new XML record:
+    logging.info('Inserting holdings record into database...')
+    holdings_table_insert.execute(
+            doi_foreign_key=doi_foreign_key,
+            xml_response=api_response_text,
+            full_text_indicator=full_text_indicator_for_doi)
+
+# =============================================================================
+# Import the list of DOIs
+# =============================================================================
+
+# Import the datset
+# Note that read_table defaults to using TSV (tab-separated) format and using
+# the first row as a header row.
+doi_dataset = pd.read_table(
+        config.input_tsv_dataset_location,
+        compression='infer')  # Allow Pandas to read compressed files.
+
+# set() will get unique values in the list:
+list_of_dois = sorted(set(
+        doi_dataset[
+                doi_dataset['oadoi_color']
```

Comment:
It would be better for this script to consume a list of DOIs rather and not hardcode this very specific structure of a single input. This change will be necessary if we query DOIs beyond the oaDOI dataset. Let's defer making a fix for later.

Reply:
I take your point. That sounds good to me.